source/common/config/delta_subscription_state.cc

```diff
@@ -0,0 +1,198 @@
+  setInitFetchTimeout(dispatcher);
+}
+
+void DeltaSubscriptionState::setInitFetchTimeout(Event::Dispatcher& dispatcher) {
+  if (init_fetch_timeout_.count() > 0 && !init_fetch_timeout_timer_) {
+    init_fetch_timeout_timer_ = dispatcher.createTimer([this]() -> void {
+      ENVOY_LOG(warn, "delta config: initial fetch timed out for {}", type_url_);
+      callbacks_.onConfigUpdateFailed(nullptr);
+    });
+    init_fetch_timeout_timer_->enableTimer(init_fetch_timeout_);
+  }
+}
+
+void DeltaSubscriptionState::pause() {
+  ENVOY_LOG(debug, "Pausing discovery requests for {}", type_url_);
+  ASSERT(!paused_);
+  paused_ = true;
+}
+
+void DeltaSubscriptionState::resume() {
+  ENVOY_LOG(debug, "Resuming discovery requests for {}", type_url_);
+  ASSERT(paused_);
+  paused_ = false;
+}
+
+// Returns true if there is any meaningful change in our subscription interest, worth reporting to
+// the server.
+void DeltaSubscriptionState::updateResourceInterest(
+    const std::set<std::string>& update_to_these_names) {
+  std::vector<std::string> cur_added;
+  std::vector<std::string> cur_removed;
+
+  std::set_difference(update_to_these_names.begin(), update_to_these_names.end(),
+                      resource_names_.begin(), resource_names_.end(),
+                      std::inserter(cur_added, cur_added.begin()));
+  std::set_difference(resource_names_.begin(), resource_names_.end(), update_to_these_names.begin(),
+                      update_to_these_names.end(), std::inserter(cur_removed, cur_removed.begin()));
+
+  for (const auto& a : cur_added) {
+    setResourceWaitingForServer(a);
+    // Removed->added requires us to keep track of it as a "new" addition, since our user may have
+    // forgotten its copy of the resource after instructing us to remove it, and so needs to be
+    // reminded of it.
+    names_removed_.erase(a);
+    names_added_.insert(a);
+  }
+  for (const auto& r : cur_removed) {
+    setLostInterestInResource(r);
+    // Ideally, when a resource is added-then-removed in between requests, we would avoid putting
+    // a superfluous "unsubscribe [resource that was never subscribed]" in the request. However,
+    // the removed-then-added case *does* need to go in the request, and due to how we accomplish
+    // that, it's difficult to distinguish remove-add-remove from add-remove (because "remove-add"
+    // has to be treated as equivalent to just "add").
+    names_added_.erase(r);
+    names_removed_.insert(r);
+  }
+}
+
+// Not having sent any requests yet counts as an "update pending" since you're supposed to resend
+// the entirety of your interest at the start of a stream, even if nothing has changed.
+bool DeltaSubscriptionState::subscriptionUpdatePending() const {
+  return !names_added_.empty() || !names_removed_.empty() ||
+         !any_request_sent_yet_in_current_stream_;
+}
+
+UpdateAck
+DeltaSubscriptionState::handleResponse(const envoy::api::v2::DeltaDiscoveryResponse& message) {
```

Comment:
Note: I haven't been considering this to be particularly callbacky, other than the fact that it's always called by a callback (onDiscoveryResponse). I think of it more as "update the state machine with this new input". If you think it makes more sense with the on..... nomenclature, though, I am fine with changing it.

Reply:
Sorry, thinking a little more, I realize I actually want DeltaSubscriptionState to be completely walled off from anything like callbacks. It should just be its own little chunk of state, with functions that modify that state and produce output. Maybe these should get even *less* callbacky names.

Reply:
This is fine, these are private methods, so all good.